groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3338 @@
+
+#include <stddef.h>
+
+// We provide two set of bsl::variant feature sets. These are the necessary
+// features for the more functional set.
+#ifdef BSLS_COMPILERFEATURES_SUPPORT_DECLTYPE
+#ifdef BSLS_COMPILERFEATURES_SUPPORT_GENERALIZED_INITIALIZERS
+#ifdef BSLS_COMPILERFEATURES_SUPPORT_RVALUE_REFERENCES
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+// Note that BSLS_LIBRARYFEATURES_HAS_CPP14_INTEGER_SEQUENCE does not require a
+// C++14 compiler
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP14_INTEGER_SEQUENCE
+#define BSL_VARIANT_FULL_IMPLEMENTATION
+#endif  // BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+#endif  // BSLS_LIBRARYFEATURES_HAS_CPP14_INTEGER_SEQUENCE
+#endif  // BSLS_COMPILERFEATURES_SUPPORT_RVALUE_REFERENCES
+#endif  // BSLS_COMPILERFEATURES_SUPPORT_GENERALIZED_INITIALIZERS
+#endif  // BSLS_COMPILERFEATURES_SUPPORT_DECLTYPE
+
+#ifdef BSL_VARIANT_FULL_IMPLEMENTATION
+#include <type_traits>
+// todo remove when replacing std::invoke with bsl::invoke
+#include <functional>
+#else  // BSL_VARIANT_FULL_IMPLEMENTATION
+#include <bslstl_function.h>
+#endif  // BSL_VARIANT_FULL_IMPLEMENTATION
+
+#ifdef BSLS_COMPILERFEATURES_SUPPORT_GENERALIZED_INITIALIZERS
+#include <initializer_list>
+#endif  // BSLS_COMPILERFEATURES_SUPPORT_GENERALIZED_INITIALIZERS
+
+#if BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
+// Include version that can be compiled with C++03 Generated on Mon Oct  3
+// 16:56:46 2022 Command line: sim_cpp11_features.pl bslstl_variant.h
+#define COMPILING_BSLSTL_VARIANT_H
+#include <bslstl_variant_cpp03.h>
+#undef COMPILING_BSLSTL_VARIANT_H
+#else
+
+namespace BloombergLP {
+namespace bslma {
+
+// A type used as a default template type argument in places where we would get
+// an empty template argument set generated by the sim_cpp11_features.pl
+// script.
+#define VARIANT_NOT_A_TYPE void
+
+#if !BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
+
+template <class TYPE = VARIANT_NOT_A_TYPE, class... OTHER>
+struct UsesBslmaAllocatorAny;
+
+template <>
+struct UsesBslmaAllocatorAny<VARIANT_NOT_A_TYPE> : bsl::false_type {
+};
+
+template <class TYPE, class... OTHER>
+struct UsesBslmaAllocatorAny
+: bsl::integral_constant<
+      bool,
+      BloombergLP::bslma::UsesBslmaAllocator<TYPE>::value ||
+          BloombergLP::bslma::UsesBslmaAllocatorAny<OTHER...>::value> {
+    // TODO : move to bslma_usesbslmaallocator or rename as
+    // Variant_UsesBslmaAllocatorAny ?
+};
+
+template <class TYPE = VARIANT_NOT_A_TYPE, class... OTHER>
+struct IsBitwiseMoveableAll;
+
+template <>
+struct IsBitwiseMoveableAll<VARIANT_NOT_A_TYPE> : bsl::false_type {
+};
+
+template <class TYPE, class... OTHER>
+struct IsBitwiseMoveableAll
+: bsl::integral_constant<bool,
+                         BloombergLP::bslmf::IsBitwiseMoveable<TYPE>::value &&
+                             IsBitwiseMoveableAll<OTHER...>::value> {
+    // TODO : move to another component or rename as
+    // Variant_IsBitwiseMoveableAll ?
```

Comment:
For these two I was wondering if it would make sense to add a generic facility that checks whether any/all of the types satisfy the given trait. They're intended to poke you in the eye so I don't forget to ask the question. :) I couldn't think of any use cases though. Can you ?

FWIW, in optional, similar utilities were in `BloombergLP::bslstl` and named `Optional_xyz`. Happy to move them to `BloombergLP::bslstlimpl` or add a comment to say they are component-private. WDYT ? Whatever we decide to do, I'll apply to all the component private utilities.

Reply:
I think we can worry about factoring them out into an independent component when the need arises.

I don't know whether BDE has an established best practice for this sort of thing, although I did coincidentally happen to hear today that there is a company-wide rule that we are not supposed to introduce subnamespaces with common names, so I guess `impl` and `detail` are out. It seems that we usually create `XImpl` or `X_Imp` classes that hold the component-private utilities for `X`, but I suppose that doesn't work in all cases; iirc, you can't have template specializations at class scope in older versions of C++.

I'll bring this up at the team meeting tomorrow and get back to you.

Reply:
I discussed this with the team and they told me that the convention is that the way you did it for `optional` is right: you put the component-private class in `bslstl` and give it a name that contains the component name and an underscore.

Nowadays many people don't know this convention, so we should also explicitly call out, for each such component-private entity, the fact that it is component-private in its docstring.

Reply:
ack, thank you for following up. Moved to `bslstl` and added documentation. Leaving  this conversation open until new changes are pushed.